lib/vmfloaty/utils.rb

```diff
@@ -46,79 +51,53 @@ def self.generate_os_hash(os_args)
     os_types
   end
 
-  def self.get_vm_info(hosts, verbose, url)
-    vms = {}
-    hosts.each do |host|
-      vm_info = Pooler.query(verbose, url, host)
-      if vm_info['ok']
-        vms[host] = {}
-        vms[host]['domain'] = vm_info[host]['domain']
-        vms[host]['template'] = vm_info[host]['template']
-        vms[host]['lifetime'] = vm_info[host]['lifetime']
-        vms[host]['running'] = vm_info[host]['running']
-        vms[host]['tags'] = vm_info[host]['tags']
-      end
-    end
-    vms
-  end
-
-  def self.prettyprint_hosts(hosts, verbose, url)
-    puts "Running VMs:"
-    vm_info = get_vm_info(hosts, verbose, url)
-    vm_info.each do |vm,info|
-      domain = info['domain']
-      template = info['template']
-      lifetime = info['lifetime']
-      running = info['running']
-      tags = info['tags'] || {}
-
-      tag_pairs = tags.map {|key,value| "#{key}: #{value}" }
-      duration = "#{running}/#{lifetime} hours"
-      metadata = [template, duration, *tag_pairs]
+  # Adapted from ActiveSupport
+  def self.strip_heredoc(str)
+    min_indent = str.scan(/^[ \t]*(?=\S)/).min
+    min_indent_size = min_indent.nil? ? 0 : min_indent.size
 
-      puts "- #{vm}.#{domain} (#{metadata.join(", ")})"
-    end
+    str.gsub(/^[ \t]{#{min_indent_size}}/, '')
   end
 
-  def self.get_all_token_vms(verbose, url, token)
-    # get vms with token
-    status = Auth.token_status(verbose, url, token)
-
-    vms = status[token]['vms']
-    if vms.nil?
-      raise "You have no running vms"
+  def self.get_service_object(type = '')
+    nspooler_strings = ['ns', 'nspooler', 'nonstandard', 'nonstandard_pooler']
+    if nspooler_strings.include? type.downcase
+      NonstandardPooler
+    else
+      Pooler
     end
-
-    running_vms = vms['running']
-    running_vms
   end
 
-  def self.prettyprint_status(status, message, pools, verbose)
-    pools.select! {|name,pool| pool['ready'] < pool['max']} if ! verbose
-
-    width = pools.keys.map(&:length).max
-    pools.each do |name,pool|
-      begin
-        max = pool['max']
-        ready = pool['ready']
-        pending = pool['pending']
-        missing = max - ready - pending
-        char = 'o'
-        puts "#{name.ljust(width)} #{(char*ready).green}#{(char*pending).yellow}#{(char*missing).red}"
-      rescue => e
-        puts "#{name.ljust(width)} #{e.red}"
+  def self.get_service_config(config, options)
```

Comment:
Hmm, I think the raedme said that the top level URL will always be used first if a services key exists? It looks like it goes for the services url if that exists and ignores the top level URL.
```
brian@biggermac:vmfloaty % cat ~/.vmfloaty.yml                                                                                   ±[●][NS-POOL]
url: "http://localhost:4567"
user: "brian"

services:
  fake:
    url: "http://asdfklajsdfkljasdflkjasdfkljasdf"
brian@biggermac:vmfloaty % floaty list --verbose                                                                                 ±[●][NS-POOL]
I, [2017-09-30T14:01:45.491612 #76248]  INFO -- : get http://asdfklajsdfkljasdflkjasdfkljasdf/vm
D, [2017-09-30T14:01:45.491692 #76248] DEBUG -- request: User-Agent: "Faraday v0.9.2"
I, [2017-09-30T14:01:45.556589 #76248]  INFO -- Status: 200
D, [2017-09-30T14:01:45.556658 #76248] DEBUG -- response: server: "nginx"
date: "Sat, 30 Sep 2017 21:01:41 GMT"
content-type: "text/html"
transfer-encoding: "chunked"
connection: "close"
vary: "Accept-Encoding"
expires: "Sat, 30 Sep 2017 21:01:40 GMT"
cache-control: "no-cache"
error: 757: unexpected token at '<html><head><meta http-equiv="refresh" content="0;url=http://webhelper.centurylink.com/index.php?origURL=http://asdfklajsdfkljasdflkjasdfkljasdf/vm&bc="/></head><body><script type="text/javascript">window.location="http://webhelper.centurylink.com/index.php?origURL="+escape(window.location)+"&r="+escape(document.referrer)+"&bc=";</script></body></html>'. Use --trace to view backtrace
```

Reply:
This was just me being unclear on the expected behavior in the readme - I updated the wording, but I originally intended it to work like this:

- Old vmfloaty config files still work as normal
- If you start defining named services in your config file, floaty will start using the first configured service by default (unless you specify a named service explicitly with `--service`)
  - If the service ends up being used is missing any keys, floaty will look at the top level values in order to attempt to fill them in, or fail if it can't find enough information.

So under this system, the case shown here is "correct", but I'm definitely open to changing how it works. Perhaps it would be more clear to continue to use the top-level values as the defaults in all cases, and only use a named service when explicitly told to do so via `--service` - what do you think?

Reply:
Ok, no I think what you detailed there in this comment is great! As long as the readme reflects that behavior I think that makes sense. I mostly wanted to protect against surprise behavior if they upgrade and suddenly their old configs start failing, but it sounds like that should be fine if they upgrade and explicitly try out this new config map, they probably intend to use that. Maybe this could be a version bump with the new config changes, so if people don't like it, they can always go back to the old vmfloaty?